plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -666,17 +712,145 @@ compileExpr e = withContextM 2 (sdToTxt $ "Compiling expr:" GHC.<+> GHC.ppr e) $
+    _                                       -> Nothing
+getSourceSpan mmb GHC.HpcTick{GHC.tickId=tid} = do
+  mb <- mmb
+  let arr = GHC.modBreaks_locs mb
+      range = Array.bounds arr
+  GHC.RealSrcSpan sp <- if Array.inRange range tid  then Just $ arr Array.! tid else Nothing
+  return sp
+getSourceSpan _ _ = Nothing
+
+-- | Obviously this function computes a GHC.RealSrcSpan from a CovLoc
+toCovLoc :: GHC.RealSrcSpan -> CovLoc
+toCovLoc sp = CovLoc (GHC.unpackFS $ GHC.srcSpanFile sp)
+                     (GHC.srcSpanStartLine sp)
+                     (GHC.srcSpanEndLine sp)
+                     (GHC.srcSpanStartCol sp)
+                     (GHC.srcSpanEndCol sp)
+
+-- Here be dragons:
+-- See Note [Tracking coverage and lazyness]
+-- See Note [Coverage order]
+-- | Annotate a term for coverage
+coverageCompile :: CompilingDefault uni fun m
+                => GHC.CoreExpr -- ^ The original expression
+                -> GHC.Type -- ^ The type of the expression
+                -> GHC.RealSrcSpan -- ^ The source location of this expression
+                -> PIRTerm uni fun -- ^ The current term (this is what we add coverage tracking to)
+                -> CoverageType -- ^ The type of coverage to do next
+                -> m (PIRTerm uni fun)
+coverageCompile originalExpr exprType src compiledTerm covT =
+  case covT of
+    -- Add a location covereage annotation to tell us "we've executed this piece of code"
+    LocationCoverage -> do
+      ann <- addLocationToCoverageIndex (toCovLoc src)
+      ty <- compileTypeNorm exprType
+      mkLazyTrace ty (T.pack . show $ ann) compiledTerm
+
+    -- Add two boolean coverage annotations to tell us "this boolean has been True/False respectively"
+    BooleanCoverage -> do
+      -- Check if the thing we are compiling is a boolean
+      CompileContext {ccNameInfo=nameInfo} <- ask
+      let bool = GHC.getName <$> Map.lookup ''Bool nameInfo -- TODO: This should always return a `Just` and it may be worthwhile to crash the compiler with an error if it doesn't
+          tyHeadName = GHC.getName <$> GHC.tyConAppTyCon_maybe exprType
+          headSymName = GHC.getName <$> findHeadSymbol originalExpr
+          isTrueOrFalse = case headSymName of
+            Nothing -> False
+            -- We have to do it this convoluted way because GHC decides that
+            -- `nm` is a *variable* but `'True` and `'False` are data constructors and
+            -- they are not equal as names even if `originalExpr` is e.g. the expression `True`
+            Just nm ->
+              or [ GHC.getOccString nm == GHC.getOccString con
+                 | Just con <- [ GHC.getName <$> Map.lookup sym nameInfo | sym <- ['True, 'False] ] -- TODO: likewise to above these two should always both return `Just`
+                 , GHC.nameSpacesRelated (GHC.occNameSpace . GHC.getOccName $ nm) (GHC.occNameSpace . GHC.getOccName $ con)
+                 ]
+      if tyHeadName /= bool || isTrueOrFalse
```

Comment:
This is all a bit weird. So you're looking for *any* boolean expression that's not a compile-time constant? I wonder if it would make sense to do it the other way around: look for places where booleans are inspected and check that you get both values. But that's just normal code coverage again: if you don't cover the case where the boolean is false, you'd be missing coverage on that branch. So... shouldn't location coverage be enough?

Reply:
That doesn't work because different boolean expressions will be squashed. Think about the code:
```
if foo x && bar y then
  <location 1>a
else
  <location 2>b
```
Here the results of `foo` and `bar` will not necessarily need to be independently covered in order to get 100% location coverage.

Furthermore, with some validators you will end up in the situation where some boolean expressions are only inspected in the `check b = if b then () else throwError ...` function. In this case relying on location coverage means you basically have to cause validation to succeed once and fail once - not ideal.